content/posts/reviewing-unsafe-rust.md

```diff
@@ -0,0 +1,593 @@
+132:        unsafe { Error::construct(error, vtable, backtrace) }
+154:        unsafe { Error::construct(error, vtable, backtrace) }
+176:        unsafe { Error::construct(error, vtable, backtrace) }
+184:    unsafe fn construct<E>(
+285:        unsafe { Error::construct(error, vtable, backtrace) }
+366:        unsafe {
+434:        unsafe {
+448:        unsafe {
+498:        unsafe {
+510:    object_drop: unsafe fn(Box<ErrorImpl<()>>),
+511:    object_ref: unsafe fn(&ErrorImpl<()>) -> &(dyn StdError + Send + Sync + 'static),
+513:    object_mut: unsafe fn(&mut ErrorImpl<()>) -> &mut (dyn StdError + Send + Sync + 'static),
+514:    object_boxed: unsafe fn(Box<ErrorImpl<()>>) -> Box<dyn StdError + Send + Sync + 'static>,
+515:    object_downcast: unsafe fn(&ErrorImpl<()>, TypeId) -> Option<NonNull<()>>,
+516:    object_drop_rest: unsafe fn(Box<ErrorImpl<()>>, TypeId),
+520:unsafe fn object_drop<E>(e: Box<ErrorImpl<()>>) {
+528:unsafe fn object_drop_front<E>(e: Box<ErrorImpl<()>>, target: TypeId) {
+538:unsafe fn object_ref<E>(e: &ErrorImpl<()>) -> &(dyn StdError + Send + Sync + 'static)
+548:unsafe fn object_mut<E>(e: &mut ErrorImpl<()>) -> &mut (dyn StdError + Send + Sync + 'static)
+557:unsafe fn object_boxed<E>(e: Box<ErrorImpl<()>>) -> Box<dyn StdError + Send + Sync + 'static>
+566:unsafe fn object_downcast<E>(e: &ErrorImpl<()>, target: TypeId) -> Option<NonNull<()>>
+583:unsafe fn context_downcast<C, E>(e: &ErrorImpl<()>, target: TypeId) -> Option<NonNull<()>>
+603:unsafe fn context_drop_rest<C, E>(e: Box<ErrorImpl<()>>, target: TypeId)
+626:unsafe fn context_chain_downcast<C>(e: &ErrorImpl<()>, target: TypeId) -> Option<NonNull<()>>
+643:unsafe fn context_chain_drop_rest<C>(e: Box<ErrorImpl<()>>, target: TypeId)
+693:        unsafe { &*(self as *const ErrorImpl<E> as *const ErrorImpl<()>) }
+701:        unsafe { &*(self.vtable.object_ref)(self) }
+708:        unsafe { &mut *(self.vtable.object_mut)(self) }
+762:        unsafe {
+```
+
+It looks like all the `unsafe` code is in `src/error.rs` and the frequent
+references to a `vtable` indicate this is probably related to some sort of
+hand-coded trait object.
+
+## Beginning the Review
+
+We'll start off the review by checking out `lib.rs`.
+
+After scrolling past almost 200 lines of top-level docs we come across our first
+interesting block of code.
+
+```rust
+// src/lib.rs
+
+mod alloc {
+    #[cfg(not(feature = "std"))]
+    extern crate alloc;
+
+    #[cfg(not(feature = "std"))]
+    pub use alloc::boxed::Box;
+
+    #[cfg(feature = "std")]
+    pub use std::boxed::Box;
+}
+```
+
+Now the code itself is quite boring, but the fact that it exists says a lot
+about this crate. Namely that they're going to great lengths to ensure `anyhow`
+is usable without the standard library.
+
+This `alloc` module allows code to use `crate::alloc::Box` when boxing things
+instead of relying on the `Box` added to the prelude by `std` (which isn't in
+the prelude for `#[no_std]` crates).
```

Comment:
FYI this isn't the main reason. We could do `extern crate alloc` unconditionally and `use alloc::boxed::Box` even in std mode, and get the same effect. But that only works on 1.36+. The more verbose setup is necessary in order for anyhow in std mode to support back to 1.34+.